[SPARK-13834][BUILD] Update sbt and sbt plugins for 2.x.

dongjoon-hyun opened the pull request:
## What changes were proposed in this pull request?

For 2.0.0, we had better make **sbt** and **sbt plugins** up-to-date. This PR checks the status of each plugins and bumps the followings.
- sbt: 0.13.9 --> 0.13.11
- sbteclipse-plugin: 2.2.0 --> 4.0.0
- sbt-dependency-graph: 0.7.4 --> 0.8.2
- sbt-mima-plugin: 0.1.6 --> 0.1.9
- sbt-revolver: 0.7.2 --> 0.8.0

All other plugins are up-to-date. (Note that `sbt-avro` seems to be change from 0.3.2 to 1.0.1, but it's not published in the repository.)

During upgrade, this PR also updated the following MiMa error. Note that the related excluding filter is already registered correctly. It seems due to the change of MiMa exception result.

```
 // SPARK-12896 Send only accumulator updates to driver, not TaskMetrics
 ProblemFilters.exclude[IncompatibleMethTypeProblem]("org.apache.spark.Accumulable.this"),
-ProblemFilters.exclude[IncompatibleMethTypeProblem]("org.apache.spark.Accumulator.this"),
+ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.Accumulator.this"),
```
## How was this patch tested?

Pass the Jenkins build.

dongjoon-hyun committed:
[SPARK-13834][BUILD] Update sbt and sbt plugins.

SparkQA commented:
**[Test build #52974 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52974/consoleFull)** for PR 11669 at commit [`d3b5d42`](https://github.com/apache/spark/commit/d3b5d423554354d935778adbd042155d8b1577d8).

SparkQA commented:
**[Test build #52979 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52979/consoleFull)** for PR 11669 at commit [`cc2bc9c`](https://github.com/apache/spark/commit/cc2bc9c7d79b844f5abeee48f8134af76189aa29).

dongjoon-hyun commented:
Hi, @JoshRosen .
Now, this PR includes SBT/MiMa updates for SPARK-13834.
Build and MiMa is tested on both JDK8 and JDK7. 
I hope it will pass at this time.

SparkQA commented:
**[Test build #52974 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52974/consoleFull)** for PR 11669 at commit [`d3b5d42`](https://github.com/apache/spark/commit/d3b5d423554354d935778adbd042155d8b1577d8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52974/
Test PASSed.

SparkQA commented:
**[Test build #52979 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52979/consoleFull)** for PR 11669 at commit [`cc2bc9c`](https://github.com/apache/spark/commit/cc2bc9c7d79b844f5abeee48f8134af76189aa29).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52979/
Test PASSed.

dongjoon-hyun commented:
Hi, @JoshRosen . 
After upgrading MiMa plugin, I can finish upgrading SBT .
Could you review and merge this? Now, it passes the test.
Thank you again for your advice at #11661 .

srowen commented:
I think that's a good idea. But are there other plugins that should be updated or can be updated too? I already updated basically all the Maven plugins, but didn't check SBT.

dongjoon-hyun commented:
Hi, @srowen . Thank you for review. 
Yes, I got inspiration for this PR from your previous JIRA issue.
I tried to upgrade `sbt` first in #11661, but it fails due to MiMa . So, I updated MiMa first and added `sbt` on that according to @JoshRosen 's advice in #11661 . For other plugins, I thought you finished already. For `sbt` and `MiMa`, I think this PR is complete for now, but I welcome any advice. 
Do you have some plugins in your mind?

srowen commented:
I don't have any in mind, and I'm not sure if there is an easy way to see if plugins have updates available, but worth a quick look while you're updating SBT and MiMa. For example `sbt-dependency-graph` is on 0.8.2 but we're using 0.7.2; `sbt-assembly` is on 0.14.2 vs 0.11.2 in the Spark build. I'm not sure if there are reasons we can't update, but I think it's worth trying it if you have some time to test. I bet there are some decent bug fixes in here.

dongjoon-hyun commented:
I've took a look quickly for plugins in `plugins.sbt` except libraries. The following are the result.

**A. Up-to-date**
- sbt-mima-plugin 0.1.9 (ok, This PR)
- sbt-idea 1.6.0  (ok)
- scalastyle-sbt-plugin 0.8.0 (ok)
- junit_xml_listener 0.5.1 (ok)
- sbt-unidoc 0.3.3 (ok)

**B. Update Needed!**
- sat-assembly  0.11.2  --> 0.14.2
- sbteclipse-plugin 2.2.0 --> 4.0.0
- sbt-dependency-graph  0.7.4 --> 0.8.2
- sbt-avro  0.3.2 --> 1.0.1
- sbt-revolver 0.7.2 --> 0.8.0

For libraries in `plubings.sbt`, there is `ANTLR`. I'm afraid that there were too much changes for that. For the plugins in list **B**, it is worthy but it might need some time to find the correct combination. That exists some version gap in some plugins. For list B, may I try in another single PR during this weekend? I can do parallel tests with Jenkins if you merge this PR. How do you think about that?

srowen commented:
Yeah, some of these plugins might be hard to update at all, or at least not to the latest version. Some may be easy and we should probably just do the easy ones. Yes this is OK by itself.

dongjoon-hyun commented:
Thank you for giving further directions, @srowen .
Then, could you merge this PR?

SparkQA commented:
**[Test build #53015 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53015/consoleFull)** for PR 11669 at commit [`99a6c76`](https://github.com/apache/spark/commit/99a6c76af3fdb4477c6f6033dddccca72f747032).

SparkQA commented:
**[Test build #53015 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53015/consoleFull)** for PR 11669 at commit [`99a6c76`](https://github.com/apache/spark/commit/99a6c76af3fdb4477c6f6033dddccca72f747032).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53015/
Test PASSed.

dongjoon-hyun commented:
Hi, @srowen . 

I updated the description of this PR and JIRA issue. 

In short, I could update most plugins except `sbt-assembly`. It show many errors in `streaming-mqtt/test:assembly`, but I cannot handle that until now. I think more experienced person can upgrade `sbt-assembly` someday soon. And, for `sbt-avro`, the current version turns out to be up-to-date. In their homepage, the 1.0 version is announced but it is not released in repository. 

Thank you again, @srowen . With your guide, this PR became more meaningful.
If you have some time, please review this again.

srowen commented on project/SparkBuild.scala:
```diff
@@ -144,7 +143,7 @@ object SparkBuild extends PomBuild {
     scalacOptions <+= target.map(t => "-P:genjavadoc:out=" + (t / "java")))
 
-  lazy val sharedSettings = graphSettings ++ sparkGenjavadocSettings ++ Seq (
+  lazy val sharedSettings = sparkGenjavadocSettings ++ Seq (
```
Was removing `graphSettings` on purpose? and the next change below. I wanted to make sure they weren't accidental.

dongjoon-hyun replied:
Sure. They are now `auto plugin`. `graphSettings` is removed now.

https://github.com/jrudolph/sbt-dependency-graph#compatibility-notes

dongjoon-hyun commented on project/SparkBuild.scala:
```diff
@@ -241,7 +240,7 @@ object SparkBuild extends PomBuild {
   (allProjects ++ optionallyEnabledProjects ++ assemblyProjects ++ Seq(spark, tools))
     .foreach(enable(sharedSettings ++ DependencyOverrides.settings ++
-      ExcludedDependencies.settings ++ Revolver.settings))
+      ExcludedDependencies.settings))
```
Here, too.

https://github.com/spray/sbt-revolver#installation

dongjoon-hyun commented:
Hi, @JoshRosen .
If it's possible, could you review this PR, too? This supersedes #11661 .
This is a build PR, so I think we need more careful review.

JoshRosen commented:
This seems fine to me. Arguably we could remove the dependency graph and revolver plugins now that they're AutoPlugins, since I imagine that a lot of people would prefer to just stick them in their global SBT plugins configuration rather than having to configure them on a per-project basis. Since other people might disagree with this, though, I think it's fine to just upgrade them and leave them here for now.

Therefore, LGTM.

dongjoon-hyun commented:
Thank you for your review, @JoshRosen !

rxin commented:
Merging in master. Thanks!

dongjoon-hyun commented:
Thank you, @rxin !
